### PR TITLE
Allow timezone-aware Cron schedules

### PR DESCRIPTION
**Pull request by tartieret**
This forks make the Cron scheduling timezone-aware. This is a feature I need and thought that could be useful to others (as it solves #57).

A new timezone field is added to the CrontabSchedule model, allowing to define for each task in which timezone the schedule should be evaluated. If no timezone information is provided, then UTC is used.

Note: this works only if Django settings USE_TZ=True

### CODE REVIEW

**Comment by tartieret** [comment]
I just added the migration file, sorry for that !

**Comment by mbaechtold** [comment]
@rightx2, you are looking at a pull request where a contributor wants the maintainers of the repository to accept some code changes. The pull request is still open, so the changes have not yet been merged into the `master` branch.

@tartieret, could you please rebase onto the `master` branch, the test suites has been fixed in https://github.com/celery/django-celery-beat/pull/118

**Comment by rightx2** [comment]
@mbaechtold Thanks for replying. Then, do I have to wait until this request is merged?...

File: requirements/default.txt [hunk 1]
```diff
@@ -1 +1,2 @@
 celery>=4.0,<5.0
+django-timezone-field==2.0
```
Comment (mbaechtold) [inline requirements/default.txt]: This seems very restrictive to me (`>=` would be appropriate if you rely on code implemented in `django-timezone-field` 2.0).

File: django_celery_beat/tzcrontab.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    def __eq__(self, other):
+        if isinstance(other, schedules.crontab):
+            return (other.month_of_year == self.month_of_year and
+                    other.day_of_month == self.day_of_month and
+                    other.day_of_week == self.day_of_week and
+                    other.hour == self.hour and
+                    other.minute == self.minute and
+                    other.tz == self.tz)
+        return NotImplemented
```
Comment (mbaechtold) [inline django_celery_beat/tzcrontab.py]: GitHub is complaining about a missing new line.

**Comment by mbaechtold** [comment]
@tartieret, could you please have a look at the test error output?

**Comment by tartieret** [comment]
Not sure why I am getting a migration conflict, I will look at it tomorrow

**Comment by onegreyonewhite** [comment]
Any updates?

**Comment by pi-bjl** [comment]
Hey @tartieret, if you are not able to continue work on this PR I would like to potentially take it over. Thanks for your work so far!

**Comment by tartieret** [comment]
Hi,

Sorry for the delay on this, I had to focus on other projects. I fixed the issues due to formatting & docs, all the test pass in Travis !

**Comment by lnagel** [comment]
@thedrow My team is very eagerly waiting for a release that contains this feature. Do you have a timeline of when a new release might happen? Thank you!!

**Comment by thedrow** [comment]
We're currently busy with the release of Celery and all other dependencies.
If you'd like to volunteer to maintain this repository we'd love that.

**Comment by auvipy** [comment]
@thedrow I am in favor off adding people as maintainers who have some track record of contribution, and really eager to dedicate their time to contribute.

**Comment by lnagel** [comment]
@thedrow @auvipy If I can help somehow, then sure, I'd love to move things along. I'm sorry for the small delay with my reply over the weekend, have been busy too with finishing off projects before heading to DjangoCon later today :)

**Comment by thedrow** [comment]
We'll take this discussion elsewhere.

**Comment by viperfx** [comment]
@tartieret Do you have timezone aware cron schedules running from the database? I am using Celery 4.2.0 and django-celery-beat@master and my tasks are not running at all.

I thought it was the timezones, so I did one hour forward, one hour back, all the combinations. Nothing.

Would be really keen to know how and if you have this configured in production?

**Comment by tartieret** [comment]
@viperfx I am not currently using it in production but it was working when I merged the pull request. Did you set your global Django timezone setting to UTC?

**Comment by viperfx** [comment]
Oh right. Do you currently use another solution for scheduling of timezone aware background tasks @tartieret 

Yes I have the default django timezone set to UTC.  Here are the timezone related settings:

```
TIME_ZONE = 'UTC'
USE_I18N = True
USE_L10N = True
USE_TZ = True
ENABLE_UTC = True
CELERY_TIMEZONE = 'UTC'
CELERY_ENABLE_UTC = True
```

**Comment by viperfx** [comment]
Afraid not. The timezones did not work for me using the beat scheduler.

What I'm doing now is using the models including the timezone field and
then using a outer task that runs every minute to run a query to check if
any tasks should run.

It works for now but I worry it might have issues down the road as the
scale increases.

Your setup seems identical, except I used redis as the broker.

**Comment by tartieret** [comment]
Here are my settings, if it helps:

Django settings.py:
```
TIME_ZONE = 'UTC'
USE_TZ = True
# Use REDIS as Result backend
CELERY_RESULT_BACKEND = "rediss://:{0}@{1}:6380/1".format(REDIS_PRIMARY, REDIS_HOST)

# Use RabbitMQ as Celery broker
CELERY_BROKER = os.environ.get('CELERY_BROKER', "amqp://guest:guest@localhost:5672/dev_vhost")
BROKER_TRANSPORT_OPTIONS = {'visibility_timeout': 3600} 
CELERY_ACCEPT_CONTENT = ['application/json']
CELERY_TASK_SERIALIZER = 'json'
```
Celery.py:
```
import os
from celery import Celery
from celery.schedules import crontab
import datetime
import pytz

...


def utc_now():
    return pytz.utc.localize(datetime.datetime.utcnow())

app = Celery('<Celery app name>')
app.now = utc_now

# Using a string here means the worker doesn't have to serialize
# the configuration object to child processes.
# - namespace='CELERY' means all celery-related configuration keys
#   should have a `CELERY_` prefix.
app.config_from_object('django.conf:settings', namespace='CELERY')

# Load task modules from all registered Django app configs.
app.autodiscover_tasks()

```

**Comment by yywing** [comment]
This not work well because Celery has a bug.

https://github.com/celery/celery/issues/5113

**Pull request merged**